### DISCUSSION TITLE
USD Support

### DISCUSSION THREAD

**Discussion by Markemp** [Markemp/Cryengine-Converter#189] [General]
Blender has stated that they plan on deprecating Collada support as early as 4.5, and definitely by the 5.0 release.  So I've started working on a USD exporter.  This is the format created by Pixar, and Blender has stated that it's going to be their preferred format in the future.

So I've been working on adding USD support.  You can track the progress under the `usd-support`  branch.  I have basic geometry and material support working, although I still have a lot to do:

- Better material support
- Armatures
- Animations

![image](https://github.com/user-attachments/assets/874943ae-4088-4ee9-b53a-b11516514ec8)

Feel free to take a look at this branch and suggest ideas/submit PRs/etc. 

I considered adding the Pixar USD project to this with a bunch of c# shims instead of creating custom serializers, but I don't have any experience using c++ code in a c# project.  If anyone has any ideas on how to implement this instead of my current method, I'd be happy to discuss it.

Thanks for reading!

**Comment by Baconator650** [comment]
This looks awesome. In regards to feedback and representing a part of the Star Citizen Community here a few points from us:

-We do materials independently in our programs (Blender/3DS Max) due to the complex node structure. Can elaborate more if interested. What we need is previously discussed is a precise naming method in order for our scripts to recognize the scripts.

-Legacy support (limited support) for Collada (In case StarFab/Blender Scripts run into unknown issues with USD)

On a personal request/research, which I kinda feel might be another discussion area;
Could you add a CGF (IVO)- CGF (Standard-ish) method for people wanting to use CryEngine to import that Star Citizen IVO formats? I still use CryEngine for a lot research and comparison to help report issues, use as fixes, export animations, and use as a general comparison. I don't know if this would increase your workload or be a good halfway step on dealing with IVO and/or could be a good method to see where issues may be coming from.

**Reply by Markemp** [reply]
The materials will be handled the same.  `<model name>_mtl_<material name>`.  I know you guys do a lot of custom, bespoke, artisanal materials so made sure to keep the same naming format. :)

I think I'd need to make a custom binary serializer to be able to export back to .cgf/.cga if I understand what you are asking.

I'm not sure what Crytek has planned for USD support.  A USD file would be most comparable to an FBX or 3DS native export.  I'll dig around to see what I can find about that, but from what is said by the Blender devs, USD seems like it'll be the way forward for most stuff.

**Reply by Baconator650** [reply]
Thank You for keeping us in mind.

Yes that was what I was thinking, as well as the other associated files (cga, cgf, chr, caf). Basically create a parallel system to test and compare assets in their native environment to one outside (ie Blender).

Unfortunately, I haven't seen anything on CryEngine's on USD support. They seem pretty set in on their custom scripts and fbx.

I wouldn't put too much work on un-ivo the files, just thought it might help with fine tunning animations and armatures. 

Out of curiosity in regards to animations, how is the output going to look like with the USD. Is CGF-Converter going to create a USDZ package to be access in the some folder as the cga (or skin)? Or is the USD with animation going to be placed in the same folder as the .caf?

**Reply by Markemp** [reply]
Ok, back to this.  A year late, but onward!

I was tentatively planning on doing the non binary format for USD (`.usda`), and including the animation data into the model file.  At least it will start out that way, and if it starts getting a bit unwieldy I'll reformat to separate animation files.  But I'll see where the project takes me.

Time to buy a ton of Anthropic API tokens.  I think I'm going to lean on Cline to do a lot of the dirty work for me. :)

**Comment by Baconator650** [comment]
Might be worth having both options depending on different users preferences or use cases.

Looking forward to it, talking with the StarFab community, it looks like we will be getting access to the animation databases.
